### PR TITLE
Implement participant unit tests

### PR DESCRIPTION
**Pull request by masih**
Implement unit tests that exhaustively assert `Participant`:

* Validates messages and their payload prior to state change.
* Verifies messages received in a variety of combinations.
* State mutation from one state to another is consistent with gpbft core algorithm.
* Handles panics gracefully from exported APIs.

The unit tests employ mocking to avoid dependency to simulation, easily create edge-cases that are hard to achieve via the existing `sim` package.

Relates to https://github.com/filecoin-project/go-f3/issues/9

### CODE REVIEW

**Review by anorth — COMMENTED** [review]
Thanks, this is a good start but I don't think it's all the testing we need at this level. The key responsibility of the Participant is to manage the transition between instances, and that flow is not yet exercised.

File: gpbft/participant_test.go [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+			t.Run("on invalid canonical chain", func(t *testing.T) {
+				host, subject := newSubject(t)
+				invalidChain := gpbft.ECChain{gpbft.ZeroTipSet()}
+				host.On("GetCanonicalChain").Return(invalidChain, *powerTable, beacon)
+				require.ErrorContains(t, subject.Start(), "invalid canonical chain")
+				host.AssertExpectations(t)
+				requireNotStarted(t, subject)
+			})
+		})
+		t.Run("on ReceiveMessage", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (masih) [reply gpbft/participant_test.go:175]: Thanks for the feedback Alex. I see two actionables:
* `RecieveMessage` on not stared participant should be error: this simplifies the execution path quite a bit, and we have the boolean to infer whther the instance is started already i.e. `granite` being `nil`.
* Test cases to compare behavour of a finished (a.k.a terminated) participant with one that's not started.

If those make sense, both of them are verbose enough that I'll approach them in follow-up PRs in favour of reducing the size of this one.

File: gpbft/participant_test.go [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+			}
+			for _, test := range tests {
+				test := test
+				t.Run(test.name, func(t *testing.T) {
+					_, subject := newSubject(t)
+					message := test.message(subject)
+					gotAccepted, gotErr := subject.ReceiveMessage(message, false)
+					require.NoError(t, gotErr)
+					require.False(t, gotAccepted)
+					gotQueued := subject.GetQueuedMessagesByInstance(message.Vote.Instance)
```
[the remark this answers is not in the document]
Comment (masih) [reply gpbft/participant_test.go:217]: That makes sense, removing deep internal state inspection.

File: gpbft/participant_test.go [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+					},
+					wantErr:     "invalid signature",
+					wantChecked: true,
+				},
+				{
+					name: "non nil Justification when not needed is error",
+					msg: &gpbft.GMessage{
+						Sender: nonZeroPowerEntry.ID,
+						Vote: gpbft.Payload{
+							Step: gpbft.PREPARE_PHASE,
```
[the remark this answers is not in the document]
Comment (masih) [reply gpbft/participant_test.go]: I have added specific cases to make the tests more comprehensive.

However they are not exhaustive. I intend to test the rest in follow up PRs, if that's OK. This PR is large enough already and is meant to focus on `Participant`, whereas the most of the logic of phase transition and quorum state is in `gpbft.go`.

File: gpbft/participant_test.go [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+						Vote: gpbft.Payload{
+							Step:  gpbft.CONVERGE_PHASE,
+							Round: 42,
+						},
+					},
+					wantErr:     "unexpected zero value for converge phase",
+					wantChecked: true,
+				},
+				{
+					name: "CONVERGE with invalid vote value is error",
```
[the remark this answers is not in the document]
Comment (masih) [reply gpbft/participant_test.go]: If I understand your comment correctly this makes the table testing pretty awkward. I have added specific cases. Please let me know if you have something specific in mind.

**Comment by masih** [comment]
Thank you for your feedback Alex, this PR is ready for another look whenever you get the chance please.

File: gpbft/participant_test.go [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.COMMIT_PHASE,
+							Value:    subject.canonicalChain,
+						},
+						Signature: signature,
+						Justification: &gpbft.Justification{
+							Vote: gpbft.Payload{
+								Instance: initialInstanceNumber + 3,
```
Comment (anorth) [inline gpbft/participant_test.go:678]: It's worth testing all possible mismatches in vote/justification here. This is a high stakes piece of code.

**Review by anorth — APPROVED** [review]

File: gpbft/participant_test.go (hunk 5 above)
Comment (masih) [reply gpbft/participant_test.go:677]: Thank you for pointing that out. 
I have (non-exhaustively) added a few more cases here. Full permutation of possible cases would need a bit more templating to reduce duplicate code. I have taken note to address this in follow up PRs as this one is verbose enough already.
